students/HarveyHerela/lesson06/mailroom4.py

```diff
@@ -0,0 +1,185 @@
+
+
+def create_report(donors):
+    """Prints a list of all donors, sorted by total amount given."""
+    report = list()
+    report.append("Donor Name    | Total Given   | Num Gifts | Average Gift")
+    report.append("\n")
+    report.append("--------------------------------------------------------")
+    report.append("\n")
+
+    # Create a list with the data to be printed
+    data_list = [create_row(name, data) for name, data in donors.items()]
+
+    # Sort the list by total amount given
+    data_list.sort(key=get_total, reverse=True)
+
+    # Define the formatter
+    line_fmtr = "{name:<14}  ${total:>13.2f}   {num:>9d}  ${avg:>12.2f}\n".format
+    
+    # Apply the formatting to all the data
+    for d in data_list:
+        report.append(line_fmtr(name=d[0], total=d[1], num=d[2], avg=d[3]))
+
+    print("\n", ''.join(report))
+
+
+def send_letters_everyone(donors):
+    """Create letters for everyone in the database, and write them to files."""
+    num = 0
+    subdir = str(datetime.now())
+    subdir = subdir.replace(':', '_')
+    try:
+        os.mkdir(subdir)
+    except OSError as oserr:
+        print(oserr)
+        print("\nThere was an error creating the directory.\n")
+        return
+        
+    # Read in the template
+    try:
+        with open("template.txt", 'r') as infile:
+            template_in = infile.read()
+    except FileNotFoundError as fnferr:
+        print(fnferr)
+        print("\nThere was an error reading the template file")
+        return
+            
+    donor_template = template_in
+    
+    for donor_name in donors:
+        # Setup up the data to write
+        donations = donors[donor_name]["donations"]
+        last = donations[-1]
+        total = sum(donations)
+        
+        # Create the personalized letter
+        # First, apply the formatting string
+        donor_template = donor_template.replace('{amount}', '{amount:.2f}')
+        donor_template = donor_template.replace('{total}', '{total:.2f}')
+        # Now, replace the keys with the data
+        try:
+            letter_out = donor_template.format(firstname=donor_name[0], 
+                lastname=donor_name[1], amount=last, total=total)
+        except KeyError as kerr:
+            print("Key Error:", kerr, "is not a valid key.",
+                "The only valid keys are: ",
+                "{firstname}, {lastname}, {amount}, and {total}")
+            return
+            
+        # Write the letter to file
+        filename = "./{dir}/{0}_{1}.txt".format(*donor_name, dir=subdir)
+        with open(filename, 'w') as outfile:
+            outfile.write(letter_out)
+            
+        num += 1
+
+    print(num, "letters created.")
+
+
+if __name__ == "__main__":
+    # Default donor list
+    donors = dict()
+    donors[("Chase", "Zuma")] = {"donations": [25]}
+    donors[("Marshall", "Rocky")] = {"donations": [30, 60]}
+    donors[("Skye", "Rubble")] = {"donations": [100, 100, 100]}
```

Comment:
I would suggest defining your donors data structure outside the `if __name__ == "__main__":` block so that you can import it into your test module.